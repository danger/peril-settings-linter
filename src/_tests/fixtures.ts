export const fixturedSettingsJSON = `
{
  "$schema": "https://raw.githubusercontent.com/danger/peril/master/peril-settings-json.schema",

  "settings": {
    "ignored_repos": ["artsy/looker", "artsy/clouds", "artsy/design", "artsy/hokusai"]
  },
  "rules": {
    "create (ref_type == tag)": "org/newRelease.ts",
    //  "pull_request": ["org/all-prs.ts", "org/jira/pr.ts"],
    "pull_request": "org/allPRs.ts",
    "pull_request.closed": "org/closedPRs.ts",
    "issues.opened": "danger/newRFC.ts",
    "issue_comment": "org/markAsMergeOnGreen.ts",
    "status.success": "org/mergeOnGreen.ts"
  },
  "repos": {
    "artsy/reaction": {
      "pull_request": "danger/pr.ts"
    },
    "artsy/force": {
      "pull_request": "dangerfile.ts"
    },
    "artsy/positron": {
      "pull_request": "dangerfile.ts"
    },
    "artsy/exchange": {
      "pull_request": "dangerfile.ts"
    },
    "artsy/volt": {
      "pull_request": "dangerfile.ts"
    }
  },
  "tasks": {
    "slack-dev-channel": "tasks/slackDevChannel.ts",
    "daily-license-check": "tasks/dailyLicenseCheck.ts",
    "rfc-summary": "tasks/weeklyRFCSummary.ts"
  },
  "scheduler": {
    "daily": "daily-license-check",
    "monday-morning-est": "rfc-summary"
  }
}
`

export const jsonSchema = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
      "GitHubInstallationSettings": {
          "properties": {
              "env_vars": {
                  "description": "An array of allowed ENV vars which are passed into Dangerfiles.",
                  "items": {
                      "type": "string"
                  },
                  "type": "array"
              },
              "ignored_repos": {
                  "description": "An array of repos that should org, and just make the occasional edge case.",
                  "items": {
                      "type": "string"
                  },
                  "type": "array"
              },
              "modules": {
                  "description": "An array of modules for Peril t the deploy, and available for Dangerfiles.",
                  "items": {
                      "type": "string"
                  },
                  "type": "array"
              }
          },
          "type": "object"
      },
      "RunnerRuleset": {
          "additionalProperties": {
              "anyOf": [
                  {
                      "items": {
                          "type": "string"
                      },
                      "type": "array"
                  },
                  {
                      "type": "string"
                  }
              ]
          },
          "type": "object"
      },
      "TaskObject": {
          "properties": {
              "daily": {
                  "type": "string"
              },
              "friday-morning-est": {
                  "type": "string"
              },
              "hourly": {
                  "type": "string"
              },
              "monday-morning-est": {
                  "type": "string"
              },
              "thursday-morning-est": {
                  "type": "string"
              },
              "tuesday-morning-est": {
                  "type": "string"
              },
              "wednesday-morning-est": {
                  "type": "string"
              },
              "weekly": {
                  "type": "string"
              }
          },
          "type": "object"
      },
      "UniqueRepoRuleset": {
          "additionalProperties": {
              "$ref": "#/definitions/RunnerRuleset"
          },
          "type": "object"
      }
  },
  "properties": {
      "repos": {
          "$ref": "#/definitions/UniqueRepoRuleset",
          "description": "A set of repos  }"
      },
      "rules": {
          "$ref": "#/definitions/RunnerRuleset",
          "description": "Having rules in here would mean that it would happen on _any_ event, anothern the DB"
      },
      "scheduler": {
          "$ref": "#/definitions/TaskObject",
          "description": " "
      },
      "settings": {
          "$ref": "#/definitions/GitHubInstallationSettings",
          "description": "In our DB this is represented as a JSON type, so you should anticipate have ."
      },
      "tasks": {
          "$ref": "#/definitions/RunnerRuleset",
          "description": "Individual tasks which a Peril can schedule, either via the Dangerfile API "
      }
  },
  "type": "object"
}
`
