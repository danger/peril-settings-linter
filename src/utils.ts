import * as octokit from "@octokit/rest"

// A bunch of this came from Danger/Peril

/**
 * Should look like one of the following:
 * - "dangerfile.js"
 * - "/path/to/dangerfile.ts"
 * - "repo/slug@to/dangerfile.ts"
 */
export type DangerfileReferenceString = string

export interface RepresentationForURL {
  /** The path the the file aka folder/file.ts  */
  dangerfilePath: string
  /** The branch to find the dangerfile on  */
  branch: string
  /** An optional repo */
  repoSlug: string | undefined
  /** The original full string, with repo etc  */
  referenceString: DangerfileReferenceString
}

/** Takes a DangerfileReferenceString and lets you know where to find it globally */
export const dangerRepresentationForPath = (value: DangerfileReferenceString): RepresentationForURL => {
  const afterAt = value.includes("@") ? value.split("@")[1] : value
  return {
    branch: value.includes("#") ? value.split("#")[1] : "master",
    dangerfilePath: afterAt.split("#")[0],
    repoSlug: value.includes("@") ? value.split("@")[0] : undefined,
    referenceString: value,
  }
}

/** Gets the contest of a file via the GH API */
export const fileContents = async (api: octokit, path: string, repoSlug: string, ref?: string) => {
  if (!repoSlug) {
    throw new Error("fileContents must include a repo in an event")
  }

  const [owner, repo] = repoSlug.split("/")
  try {
    const response = await api.repos.getContent({ repo, owner, path, ref })
    if (response && response.data && response.data.type === "file") {
      const buffer = Buffer.from(response.data.content, response.data.encoding)
      return buffer.toString()
    } else {
      return ""
    }
  } catch {
    return ""
  }
}

// Taken from https://gist.github.com/zenparsing/5dffde82d9acef19e43c
export function dedent(callSite, ...args) {
  function format(str) {
    let size = -1

    return str.replace(/\n(\s+)/g, (m, m1) => {
      if (size < 0) {
        size = m1.replace(/\t/g, "    ").length
      }

      return "\n" + m1.slice(Math.min(m1.length, size))
    })
  }

  if (typeof callSite === "string") {
    return format(callSite)
  }

  if (typeof callSite === "function") {
    return (...aargs) => format(callSite(...aargs))
  }

  const output = callSite
    .slice(0, args.length + 1)
    .map((text, i) => (i === 0 ? "" : args[i - 1]) + text)
    .join("")

  return format(output)
}
