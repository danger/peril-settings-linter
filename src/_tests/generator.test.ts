import { fileMapForPerilSettingsRepo } from "../"
import * as utils from "../utils"

// @ts-ignore
utils.fileContents = jest.fn()

import { fixturedPerilRunnerDeps } from "./fixtures"

it("creates the right typescript setup", async () => {
  const mockContent: jest.Mock<string> = utils.fileContents as any
  // 1st = peril runner package json
  mockContent.mockReturnValueOnce(Promise.resolve(fixturedPerilRunnerDeps))

  const api = {} as any // NOOP because of the above mock

  const map = await fileMapForPerilSettingsRepo(api, {
    isPublic: true,
    setupTests: true,
    useTypeScript: true,
    repo: {
      name: "Fake Repo",
      owner: {
        login: "User Name",
      },
    },
  })

  expect(map).toMatchSnapshot()
})

it("creates a JS with no tests setup", async () => {
  const mockContent: jest.Mock<string> = utils.fileContents as any
  // 1st = peril runner package json
  mockContent.mockReturnValueOnce(Promise.resolve(fixturedPerilRunnerDeps))

  const api = {} as any // NOOP because of the above mock

  const map = await fileMapForPerilSettingsRepo(api, {
    isPublic: false,
    setupTests: false,
    useTypeScript: false,
    repo: {
      name: "Fake Repo",
      owner: {
        login: "User Name",
      },
    },
  })

  expect(map).toMatchSnapshot()
})
