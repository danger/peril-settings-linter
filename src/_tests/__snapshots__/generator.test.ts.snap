// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a JS with no tests setup 1`] = `
Object {
  ".gitignore": "
logs
* .log
npm - debug.log *
yarn - debug.log *

# Dependency directories
node_modules
jspm_packages

# Optional npm cache directory
.npm

# Optional REPL history
.node_repl_history
",
  "README.md": "
## User Name Peril Settings

### What is this project?

This is the configuration repo for Peril on the Artsy org. There is a [settings file](settings.json) and org-wide
dangerfiles which are inside the [org folder](org/).

Here's some links to the key concepts:

- [Peril](https://github.com/danger/peril)
- [Danger JS](http://danger.systems/js/)
- [Peril for Orgs](https://github.com/danger/peril/blob/master/docs/setup_for_org.md)
- [Introducing Peril to the Artsy org](http://artsy.github.io/blog/2017/09/04/Introducing-Peril/)
- [Peril's Dashboard 🔐](https://staging-dashboard.peril.systems/)

### TL:DR on this Repo?

Peril is Danger running on a web-server, this repo is the configuration for that. The dangerfiles for our org live
in [\`org\`](org)

### To Develop

\`\`\`sh
git clone https://github.com/User Name/Fake Repo.git
cd Fake Repo
yarn install
code .
\`\`\`

You will need node and yarn installed beforehand. You can get them both by running \`brew install yarn\`. This will give
you auto-completion and types for Danger/Peril when working on your rules.

#### Adding a rule

Create a new rule by making a new Dangerfile:


\`\`\`ts
// org/myRuleName.js
export default async () => {
  // [...]
})
\`\`\`


",
  "org/checkForPRBody.js": "
export default () => {
const pr = danger.github.pr
if (pr.body === null || pr.body.length === 0) {
  fail(\\"Please add a description to your PR.\\")
}
}
",
  "package.json": "{
  \\"scripts\\": {},
  \\"dependencies\\": {
    \\"@octokit/rest\\": \\"16.15.0\\",
    \\"@sentry/node\\": \\"4.1.1\\",
    \\"@slack/client\\": \\"4.8.0\\",
    \\"agenda\\": \\"1.0.3\\",
    \\"apollo-server-express\\": \\"1.4.0\\",
    \\"async-exit-hook\\": \\"2.0.1\\",
    \\"aws-sdk\\": \\"2.374.0\\",
    \\"babel-polyfill\\": \\"7.0.0-alpha.19\\",
    \\"body-parser\\": \\"1.18.3\\",
    \\"chalk\\": \\"2.4.2\\",
    \\"cookie\\": \\"0.3.1\\",
    \\"cookie-parser\\": \\"1.4.3\\",
    \\"cors\\": \\"2.8.4\\",
    \\"danger\\": \\"7.0.13\\",
    \\"debug\\": \\"4.1.1\\",
    \\"dotenv\\": \\"5.0.1\\",
    \\"express\\": \\"4.16.4\\",
    \\"express-x-hub\\": \\"1.0.4\\",
    \\"github-webhook-event-types\\": \\"1.2.1\\",
    \\"graphql\\": \\"0.13.2\\",
    \\"graphql-playground-middleware-express\\": \\"1.7.5\\",
    \\"graphql-relay\\": \\"0.5.5\\",
    \\"graphql-resolvers\\": \\"0.2.2\\",
    \\"graphql-tools\\": \\"3.1.1\\",
    \\"graphql-tools-types\\": \\"1.1.26\\",
    \\"json5\\": \\"2.1.0\\",
    \\"jsonwebtoken\\": \\"8.4.0\\",
    \\"lodash\\": \\"4.17.11\\",
    \\"mongoose\\": \\"5.3.4\\",
    \\"node-fetch\\": \\"2.3.0\\",
    \\"node-mocks-http\\": \\"1.7.3\\",
    \\"override-require\\": \\"1.1.1\\",
    \\"primus\\": \\"7.2.3\\",
    \\"require-from-string\\": \\"2.0.2\\",
    \\"url\\": \\"0.10.3\\",
    \\"uuid\\": \\"3.3.2\\",
    \\"winston\\": \\"3.2.1\\",
    \\"winston-papertrail\\": \\"1.0.5\\"
  },
  \\"devDependencies\\": {
    \\"@graphql-inspector/core\\": \\"0.13.2\\",
    \\"@babel/core\\": \\"7.3.3\\",
    \\"@babel/plugin-proposal-class-properties\\": \\"7.3.3\\",
    \\"@babel/plugin-proposal-object-rest-spread\\": \\"7.3.2\\",
    \\"@babel/plugin-transform-flow-strip-types\\": \\"7.2.3\\",
    \\"@babel/plugin-transform-regenerator\\": \\"7.0.0\\",
    \\"@babel/preset-env\\": \\"7.3.1\\",
    \\"@babel/preset-typescript\\": \\"7.3.3\\",
    \\"@types/lodash\\": \\"^4.14.116\\",
    \\"@types/node\\": \\"^9.1.2\\",
    \\"@types/node-fetch\\": \\"^2.1.2\\",
    \\"danger-plugin-spellcheck\\": \\"^1.2.3\\",
    \\"danger-plugin-yarn\\": \\"^1.2.1\\",
    \\"googleapis\\": \\"^36.0.0\\",
    \\"graphql-relay-tools\\": \\"0.1.1\\",
    \\"graphql-schema-utils\\": \\"0.6.6\\",
    \\"jira-client\\": \\"6.4.1\\",
    \\"semver-sort\\": \\"0.0.4\\"
  }
}",
  "peril.settings.json": "{
  \\"$schema\\": \\"https://raw.githubusercontent.com/danger/peril/master/peril-settings-json.schema\\",
  \\"rules\\": {
    \\"pull_request\\": \\"org/checkForPRBody.js\\"
  }
}",
}
`;

exports[`creates the right typescript setup 1`] = `
Object {
  ".gitignore": "
logs
* .log
npm - debug.log *
yarn - debug.log *

# Dependency directories
node_modules
jspm_packages

# Optional npm cache directory
.npm

# Optional REPL history
.node_repl_history
",
  "LICENSE": "The MIT License (MIT)

Copyright (c) User Name

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the \\"Software\\"), to deal
    in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do { so, subject } to the following conditions:

    The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED \\"AS IS\\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
    ",
  "README.md": "
## User Name Peril Settings

### What is this project?

This is the configuration repo for Peril on the Artsy org. There is a [settings file](settings.json) and org-wide
dangerfiles which are inside the [org folder](org/).

Here's some links to the key concepts:

- [Peril](https://github.com/danger/peril)
- [Danger JS](http://danger.systems/js/)
- [Peril for Orgs](https://github.com/danger/peril/blob/master/docs/setup_for_org.md)
- [Introducing Peril to the Artsy org](http://artsy.github.io/blog/2017/09/04/Introducing-Peril/)
- [Peril's Dashboard 🔐](https://staging-dashboard.peril.systems/)

### TL:DR on this Repo?

Peril is Danger running on a web-server, this repo is the configuration for that. The dangerfiles for our org live
in [\`org\`](org)

### To Develop

\`\`\`sh
git clone https://github.com/User Name/Fake Repo.git
cd Fake Repo
yarn install
yarn jest

code .
\`\`\`

You will need node and yarn installed beforehand. You can get them both by running \`brew install yarn\`. This will give
you auto-completion and types for Danger/Peril when working on your rules.

#### Adding a rule

Create a new rule by making a new Dangerfile:


\`\`\`ts
// org/myRuleName.ts
export default async () => {
  // [...]
})
\`\`\`


#### Testing a rule

We use Jest to test our Dangerfiles. It uses the same techniques as testing a
[danger plugin](http://danger.systems/js/usage/extending-danger.html) where the global imports from danger are faked.

1.  Create a file for your rule: \`org/_tests/[x].test..ts\`.
2.  Add a \`before\` and \`after\` setting up and resetting mocks:

    \`\`\`ts
    jest.mock(\\"danger\\", () => jest.fn())
    import * as danger from \\"danger\\"
    const dm = danger as any

    beforeEach(() => {
      dm.danger = {}
      dm.fail = jest.fn() // as necessary
    })

    afterEach(() => {
      dm.fail = undefined
    })
    \`\`\`

3.  Set up your danger object and run the function exported in \`all-prs.ts\`:

    \`\`\`ts
    import rfcN from \\"../org/all-prs\\"

    it(\\"warns when there's there's no assignee and no WIP in the title\\", async () => {
      dm.danger.github = { pr: { title: \\"Changes to the setup script\\", assignee: null } }
      await rfcN()

      expect(something).toHappen()
        // [...]
      })
    })
    \`\`\`

4.  Validate that the \`fail\`/\`warn\`/\`message\`/\`markdown\` functions are called.


",
  "org/checkForPRBody.ts": "
export default () => {
const pr = danger.github.pr
if (pr.body === null || pr.body.length === 0) {
  fail(\\"Please add a description to your PR.\\")
}
}
",
  "package.json": "{
  \\"scripts\\": {
    \\"type-check\\": \\"tsc --no-emit\\",
    \\"test\\": \\"jest\\"
  },
  \\"license\\": \\"MIT\\",
  \\"dependencies\\": {
    \\"@octokit/rest\\": \\"16.15.0\\",
    \\"@sentry/node\\": \\"4.1.1\\",
    \\"@slack/client\\": \\"4.8.0\\",
    \\"agenda\\": \\"1.0.3\\",
    \\"apollo-server-express\\": \\"1.4.0\\",
    \\"async-exit-hook\\": \\"2.0.1\\",
    \\"aws-sdk\\": \\"2.374.0\\",
    \\"babel-polyfill\\": \\"7.0.0-alpha.19\\",
    \\"body-parser\\": \\"1.18.3\\",
    \\"chalk\\": \\"2.4.2\\",
    \\"cookie\\": \\"0.3.1\\",
    \\"cookie-parser\\": \\"1.4.3\\",
    \\"cors\\": \\"2.8.4\\",
    \\"danger\\": \\"7.0.13\\",
    \\"debug\\": \\"4.1.1\\",
    \\"dotenv\\": \\"5.0.1\\",
    \\"express\\": \\"4.16.4\\",
    \\"express-x-hub\\": \\"1.0.4\\",
    \\"github-webhook-event-types\\": \\"1.2.1\\",
    \\"graphql\\": \\"0.13.2\\",
    \\"graphql-playground-middleware-express\\": \\"1.7.5\\",
    \\"graphql-relay\\": \\"0.5.5\\",
    \\"graphql-resolvers\\": \\"0.2.2\\",
    \\"graphql-tools\\": \\"3.1.1\\",
    \\"graphql-tools-types\\": \\"1.1.26\\",
    \\"json5\\": \\"2.1.0\\",
    \\"jsonwebtoken\\": \\"8.4.0\\",
    \\"lodash\\": \\"4.17.11\\",
    \\"mongoose\\": \\"5.3.4\\",
    \\"node-fetch\\": \\"2.3.0\\",
    \\"node-mocks-http\\": \\"1.7.3\\",
    \\"override-require\\": \\"1.1.1\\",
    \\"primus\\": \\"7.2.3\\",
    \\"require-from-string\\": \\"2.0.2\\",
    \\"url\\": \\"0.10.3\\",
    \\"uuid\\": \\"3.3.2\\",
    \\"winston\\": \\"3.2.1\\",
    \\"winston-papertrail\\": \\"1.0.5\\"
  },
  \\"devDependencies\\": {
    \\"@graphql-inspector/core\\": \\"0.13.2\\",
    \\"@babel/core\\": \\"7.3.3\\",
    \\"@babel/plugin-proposal-class-properties\\": \\"7.3.3\\",
    \\"@babel/plugin-proposal-object-rest-spread\\": \\"7.3.2\\",
    \\"@babel/plugin-transform-flow-strip-types\\": \\"7.2.3\\",
    \\"@babel/plugin-transform-regenerator\\": \\"7.0.0\\",
    \\"@babel/preset-env\\": \\"7.3.1\\",
    \\"@babel/preset-typescript\\": \\"7.3.3\\",
    \\"@types/lodash\\": \\"^4.14.116\\",
    \\"@types/node\\": \\"^9.1.2\\",
    \\"@types/node-fetch\\": \\"^2.1.2\\",
    \\"danger-plugin-spellcheck\\": \\"^1.2.3\\",
    \\"danger-plugin-yarn\\": \\"^1.2.1\\",
    \\"googleapis\\": \\"^36.0.0\\",
    \\"graphql-relay-tools\\": \\"0.1.1\\",
    \\"graphql-schema-utils\\": \\"0.6.6\\",
    \\"jira-client\\": \\"6.4.1\\",
    \\"semver-sort\\": \\"0.0.4\\",
    \\"typescript\\": \\"*\\",
    \\"jest\\": \\"*\\"
  }
}",
  "peril.settings.json": "{
  \\"$schema\\": \\"https://raw.githubusercontent.com/danger/peril/master/peril-settings-json.schema\\",
  \\"rules\\": {
    \\"pull_request\\": \\"org/checkForPRBody.ts\\"
  }
}",
  "tsconfig.json": "{
  \\"compilerOptions\\": {
    \\"target\\": \\"es2017\\",
    \\"module\\": \\"commonjs\\",
    \\"resolveJsonModule\\": true,
    \\"allowSyntheticDefaultImports\\": true,
    \\"lib\\": [
      \\"es2017\\"
    ],
    \\"strict\\": true
  }
}",
}
`;
